### ISSUE TITLE
bosco install / cdn with arbitrary node versions

### ISSUE THREAD

**Issue by cliftonc** [tes/bosco#127]
We've now well and truly gone down the route of running services in whatever node version we like, and at TES have everything from v0.10.43 to v5.3.0 and everything in between.

While this may be the true embodiment of micro-services (yay!), it creates a real problem for the tooling, esp. the commands that depend intrinsically on the node version the service expects:

- [X] run / stop : this is handled by pm2
- [x] cdn : not handled, crucial because of (ever more complex) build steps
- [x] install : not handled

Both ```cdn``` and ```install``` typically use npm wrappers, so the task is actually to run the task using the version of npm that the service expects (note that some services use gulp directly so this isn't always true - lets park this, and assume everyone moves to npm scripts).

How do you programmatically select the npm version to use?

### Attempt 1: Copy Node Runner

We can use the same approach to retrieve the interpreter that we use in the Node run wrapper (at least its consistent), and append that to the beginning of the build / watch commands.

So it would resolve:

``` /Users/ccunningham/.nvm/versions/node/v5.3.0/bin/ ```

We then append the build command:  ```npm run build-assets```

```/Users/ccunningham/.nvm/versions/node/v5.3.0/bin/npm run build-assets```

If you run the command like this however, it then immediately ends up calling out to other globally installed commands like babel-node, e.g.

```
> app-home@2.0.0 build-assets /Users/ccunningham/work/404/app-home
> babel-node ./node_modules/gulp/bin/gulp.js build
```

And then failing.

### Attempt 2: Use nvm

As a hacky test to see if NVM would work, I simply appended ```nvm use; ``` to the front of the build command in bosco-service.json.  This will assume that every service has an .nvmrc.  

```nvm use; run build-assets```

This works (in that it builds), but it also sets your **global** nvm version to whatever the last service would have used.

This could technically be overcome (given they're all serially built), but storing what it is before running the command, and then setting it to that afterwards.  We would need to add an .nvmrc to every service to ensure that we know what service it is actually running on.

@Fetz suggested something similar in techy chat:

```$NVM_DIR/nvm.sh run "$NVM_BIN/npm run build-assets\"```

This requires that you have completely followed the NVM install steps:  https://github.com/creationix/nvm#manual-install

### Option 3: Run everything on node 5

Just upgrade, and run everything on node 5.  Some stuff breaks (as not all modules work in this brave new world).

### Option 4: Run everything on node 4

Work backwards, stay on 4 until we are ready to upgrade everything to 5.

### Option 5:  Run non team services as docker containers

Rather than including other node projects (like app-authentication) in your team, and having to run it 'locally', we could have them run as docker containers like the infra-* projects.

This could quite easily be an extension of the way the dependency resolution works now, where it goes off and gets the info about the service from github and tries to pull and start the matching container.  The biggest challenge is mapping all the ports in VirtualBox to make it work, but @cressie176 thinks this could potentially be solved by linking them into infra-nginx-gateway and routing everything from there.

**Comment by cliftonc** [comment]
I'll say that my strong preference at this stage is to go back to one agreed version of Node that we regularly bump forward en-masse, mostly due to the fact that I can't see a way of making this work that isn't a complete pain in the a** and insanely confusing / complex for those working across teams / services.

:+1: for Option 4

**Comment by dmorgantini** [comment]
@bettiolo has a mechanism using local.tes-dev.com that would allow you to not need registration stuff locally (ie: you can use dev cookies) but I haven't seen any documentation on how to actually set that up.

**Comment by geophree** [comment]
I'm happy with forcing everyone to have a correctly installed `nvm`.  We assume it's in `~/.nvm` in `src/RunWrappers/Node.js:getInterpreter:53-54`.  We may want to change that to use `$NVM_DIR` though.

Also, this command probably isn't going to work as-typed `$NVM_DIR/nvm.sh run "$NVM_BIN/npm run build-assets"` because `$NVM_BIN` will be replaced too soon and therefore with the wrong version.

I think you were on the right track with `nvm use; npm run script build-assets`, we just need to put it in a subshell: `bash -c 'nvm use; npm run build-assets'` so the only thing it does that will affect the global state is to change the `/current` symlink in the nvm directory.  We can easily switch it back afterwards because we have the old state in our process.

**Comment by madeleinee** [comment]
Hi did you mean to cc me

Product Director, Times Higher Education

On 9 Mar 2016, at 20:18, Jeff Hanke <notifications@github.com<mailto:notifications@github.com>> wrote:


I'm happy with forcing everyone to have a correctly installed nvm. We assume it's in ~/.nvm in src/RunWrappers/Node.js:getInterpreter:53-54. We may want to change that to use $NVM_DIR though.

Also, this command probably isn't going to work as-typed $NVM_DIR/nvm.sh run "$NVM_BIN/npm run build-assets" because $NVM_BIN will be replaced too soon and therefore with the wrong version.

I think you were on the right track with nvm use; npm run script build-assets, we just need to put it in a subshell: bash -c 'nvm use; npm run build-assets' so the only thing it does that will affect the global state is to change the /current symlink in the nvm directory. We can easily switch it back afterwards because we have the old state in our process.

—
Reply to this email directly or view it on GitHub<https://github.com/tes/bosco/issues/127#issuecomment-194489002>.
This e-mail message, including any attachments transmitted with it, is CONFIDENTIAL and may contain legally privileged or private information. This message is intended solely for the use of the individual or entity to whom it is addressed. 

 

If you have received this message in error, please notify the sender or security@tesglobal.com immediately and delete it from your system. You should not copy, store or use this email or any attachment(s) for any purpose, nor disclose their contents to any other person. 

 

E-mail is not a 100% virus-free or secure medium. Whilst TSL scans all communications for viruses or other malicious software, we do not guarantee that this message is virus-free and accept no liability for any potential loss or damage that may arise as a result of receiving, storing and reading this message and any attachment(s).  It is your responsibility to ensure that viruses do not adversely affect your system and that your messages to us meet your own security requirements. 

 

We reserve the right to read any e-mail or attachment entering or leaving our systems without notice.

 

TSL Education is a limited company registered in England and Wales (Company number 02017289) with its registered office at 26 Red Lion Square, London, WC1R 4HQ.

**Comment by cressie176** [comment]
My vote goes to node 4.2.0 It's the most simple for now. Interested in @bettiolo's solution though, and how this is configured - e.g. if you want a local page composer but a development site-assets how would this work?

**Comment by Fetz** [comment]
@geophree I tried to use (`bosco cdn`)  `nvm use; npm run script build-assets` and didn't work (with current implementation of `bosco cdn`) and with this `$NVM_DIR/nvm.sh run "$NVM_BIN/npm run build-assets` it worked.
The only thing that I need to confirm is that executing `nvm run` if I need to set the version like `nvm run <node_version> <task>` (maybe he is using the default version) or nvm will detect the version correctly.

I think that the `nvm use; npm run script build-assets` doesn't work with current implementation is that `nvm use` will look for the current directory to check what node should use, so we may need to before running the script to change to the folder first `cd <folder_script_path>; nvm use; npm run script build-assets;`

Regarding @bettiolo I assume that is regarding the pull request for compoxure app. If yes, that was done for this hackaton app [extension-compoxure-chrome](https://github.com/tes/extension-compoxure-chrome) so we can test our local app (compoxure markup) with different environments. The prove of concept is working. But will need some improvements to be fully functional.

Another thing I used when I was having some issues with nginx and integrating the jobs with registration (thanks again @cressie176 ) was to override resources in different environments using [resource-override](https://chrome.google.com/webstore/detail/resource-override/pkoacgokdfckfpndoffpifphamojphii?hl=en) so you can test your css and js without releasing it (eg: using your local js in staging)

**Comment by cliftonc** [comment]
The CDN cmd always executes in the folder of the service, that's already taken care of.  Will try Jeff's trick today.

**Comment by cliftonc** [comment]
Using ```bash -c 'nvm use; npm run build-assets'``` doesn't work for me because I don't use bash, and so don't have nvm installed (I think).  My bash fu isn't great so need someone else to help.

**Comment by sithmel** [comment]
have you tried ```/bin/sh -c``` ?

**Comment by cliftonc** [comment]
I just get:

```
ccunningham@LT4593  ~/work/404/app-home   master ●  ~/work/bosco/bin/bosco.js cdn -s                                                                                        130 ↵  5192  09:27:28
[09:27:38] Bosco: Error requiring command file: /Users/ccunningham/work/bosco/commands/audit.js: Error: Cannot find module 'nsp/lib/check'
[09:27:38] Bosco: Freeze! Peace officers!
[09:27:38] Bosco: Initialised using [/Users/ccunningham/.bosco/bosco.json] in environment [local] with team [tes/404]
[09:27:38] Bosco: Starting pseudo CDN on port: 7334
'/bin/sh -c \'nvm use; npm run build-assets\''
[09:27:38] Bosco: Running build command for app-home: npm run build-assets with /bin/sh -c
[09:27:40] Bosco: Failed build command for app-home exited with code 1:

> app-home@2.0.0 build-assets /Users/ccunningham/work/404/app-home
> babel-node ./node_modules/gulp/bin/gulp.js build

[09:27:40] Bosco: /bin/sh: nvm: command not found
```

**Comment by cliftonc** [comment]
It looks like it is running the build-assets command before nvm use is finished (is it because it shells off somewhere else?)

**Comment by BorePlusPlus** [comment]
AFAIK nvm is provided by a function that needs to be sourced in the shell you want to use it from.

**Comment by codeaholics** [comment]
Correct. Otherwise it wouldn't be able to affect shell state such as the PATH.

**Comment by cliftonc** [comment]
All very helpful @codeaholics and @BorePlusPlus input, what do you propose I do instead?

**Comment by glasseyes42** [comment]
if nvm is missing - it suggests the profile is not being picked up right? Running a bash shell with the right env variables (especially HOME) might work?

**Comment by codeaholics** [comment]
```bash -c 'source $HOME/.nvm/nvm.sh ; nvm use ; npm run build-assets'```

**Comment by cliftonc** [comment]
Ok .. seemed to work, but the shell swallowed the output - any way I can see it?

**Comment by glasseyes42** [comment]
I thought you could specify a shell and env vars in the child process creation?

**Comment by bettiolo** [comment]
I vote for Option 2 and Option 6:

Option 6: Use non team services from DEV

TLDR;
- Depend on APIs from development
- Depend on composure backends from development (except for the code of your own app)

Longer:
I don't really like to have ultrabytes of memory used up by a lot of dependencies running on my laptop. What I do is that I point my local apps to use my dependencies from development.tescloud.com.

To get the cookies working locally, while avoiding to spin up all the registration bits, I use a record that I got setup: local.tes-dev.com. This resolves to localhost so I can use cookies issued in development while hitting the app that I am running locally.

This setup makes developing api locally very simple, no bosco needed, no complex managing of dependent services.

It gets more complicated when running through page composer as it will try to resolve things locally. We built the https://github.com/tes/extension-compoxure-chrome during the hackathon that allows you to parse a compoxure template from your local service and get it resolved in development, staging and live. It is still rough.

Another useful functionality is that you can configure page-composer to get resources from another environment: https://github.com/tes/service-page-composer#pegging-local-references-of-services-to-other-environments

Plus, I use consistent ports between all the environments, this allows me to use the same configuration for LOCAL and DEV, I dropped LOCAL.

The added side effect is that running my dependencies over network makes it more clear where I have performance issues as the latency will be much higher VS the loopback kernel interface

**Comment by cliftonc** [comment]
Got the use of .nvmrc to work on both install and CDN:

https://github.com/tes/bosco/pull/128

Issues with some tests however (may need some help on these @geophree).

**Comment by cliftonc** [comment]
This is resolved with the PR from #128
